presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -85,15 +90,15 @@
     protected final HiveMetastore delegate;
     private final LoadingCache<String, Optional<Database>> databaseCache;
     private final LoadingCache<String, List<String>> databaseNamesCache;
-    private final LoadingCache<HiveTableName, Optional<Table>> tableCache;
+    private final LoadingCache<HiveTableNameWithContext, Optional<Table>> tableCache;
```

Comment:
Instead of adding several `HiveTableNameWithContext` classes, you can have single `WithContext<T>`.
Something like this:

```java
class WithContext<T>
{
    private final HiveContext context;
    private final T key;

    public WithContext(HiveContext context, T key)
    {
        this.context = requireNonNull(context, "context is null");
        this.key = requireNonNull(key, "key is null");
    }

    public HiveContext getContext()
    {
        return context;
    }

    public T getKey()
    {
        return key;
    }

    @Override
    public boolean equals(Object o)
    {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        WithContext<?> other = (WithContext<?>) o;
        return Objects.equals(context, other.context) &&
                Objects.equals(key, other.key);
    }

    @Override
    public int hashCode()
    {
        return Objects.hash(context, key);
    }

    @Override
    public String toString()
    {
        return toStringHelper(this)
                .add("context", context)
                .add("key", key)
                .toString();
    }
}
```

Reply:
One complication is that this needs to work with JSON serialization for `RecordingHiveMetastore`, so we'd need to enumerate the subclasses with `@JsonSubTypes`.

Reply:
@electrum note that `RecordingHiveMetastore` was not touched in this PR (at least as of yesterday), so it doesn't use the new classes introduced here. (So no serialization problem)
I think this is actually good enough, given the purpose the `RecordingHiveMetastore` has.